app/sender.go

```diff
@@ -0,0 +1,110 @@
+package app
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	avro "github.com/topfreegames/avro/go/eventsgateway/generated"
+	"github.com/topfreegames/eventsgateway/forwarder"
+	pb "github.com/topfreegames/protos/eventsgateway/grpc/generated"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type Sender interface {
+	SendEvents(msg *pb.Events) (*pb.Response, error)
+	SendEvent(msg *pb.Event) (*pb.Response, error)
+}
+
+type sender struct {
+	config      *viper.Viper
+	logger      logrus.FieldLogger
+	producer    forwarder.Forwarder
+	topicPrefix string
+}
+
+func NewSender(
+	producer forwarder.Forwarder,
+	logger logrus.FieldLogger,
+	config *viper.Viper,
+) Sender {
+	s := &sender{producer: producer, logger: logger, config: config}
+	s.configure()
+	return s
+}
+
+func (s *sender) configure() {
+	s.loadConfigurationDefaults()
+	s.topicPrefix = s.config.GetString("extensions.kafkaproducer.topicPrefix")
+}
+
+func (s *sender) loadConfigurationDefaults() {
+	s.config.SetDefault("extensions.kafkaproducer.topicPrefix", "sv-uploads-")
+}
+
+// SendEvents sends a batch of events to kafka
+func (s *sender) SendEvents(msg *pb.Events) (*pb.Response, error) {
+	for i := range msg.Events {
+		if _, err := s.SendEvent(msg.Events[i]); err != nil {
+			return &pb.Response{FailureIndex: int64(i)}, err
```

Comment:
If several messages fail we'll only now the last index, right? Shouldn't we keep an array of errored indexes?

Reply:
The server returns after first fail and the client will retry with a slice from that index on.

Reply:
You're correct, I misread the loop. Should we stop at the first error or retry only errored ones? I we are sending 10 evts and the first one is malformed even though we retry it several times we'll never send the following 9 events

Reply:
I think it's simpler to reason about the code this way. I'm not sure there's a gain in keeping track of all failures. When there's an outage probably every SendEvent will fail or most of them and we would have to retry sending a large percentage of the original array anyway.
Since we're logging retries and indexes, we can check how many will fail and at which index.